Convert CorrelationAnalysis to a class

josephmure opened the pull request:
Upgrade documentation (API, theory, example).
Fixes #1753

## Notable API changes

- `SRC` becomes `computeSquaredSRC`.
- `SignedSRC` becomes `computeSRC`.
- `SRRC` is dropped: it called `SRC` (which computed the *squared* SRC) on the ranked samples.
- A new `computeSRRC` method is created: it calls `computeSRC` on the ranked samples.

## Following Vincent's review https://github.com/openturns/openturns/pull/2056#issuecomment-1141906413

- Create a new `CorrelationAnalysis::computeKendallTau` method based on the `CorrelationAnalysis::computePearsonCorrelation` blueprint.

jschueller commented on lib/src/Base/Stat/openturns/CorrelationAnalysis.hxx:
```diff
@@ -34,37 +35,69 @@ BEGIN_NAMESPACE_OPENTURNS
+  String __repr__() const override;
+
+  /** Compute the Pearson correlation coefficient between the input sample and the output sample */
+  const Point computePearsonCorrelation();
```
avoid const for return types when its not a pointer

josephmure replied:
Oh yes, that's what I was trying to do.

josephmure replied:
Fixed.

jschueller commented:
there should be backward-compatibility paths

josephmure commented:
@jschueller This PR does factorize some code (but not much, I agree):

- the dimension checks at the start of each of the old functions are now done in the constructor only,
- the code of the old "SRC" function is considerably simplified in the new "computeSquaredSRC" function because it now relies on "computeSRC" instead of rewriting the same lines of code.

Concerning backward compatibility, I believe @mbaudin47 argued in favor of breaking compatibility in order to make users realize that "computeSRC" does not do what "SRC" previously did. I don't think we reached a consensus on the subject.

regislebrun commented:
According to the minutes of the meeting, we decided to perform the following changes:
+ ComputeSRC->ComputeSquaredSRC
+ ComputeSignedSRC->ComputeSRC

josephmure commented:
@regislebrun  the original names were simply "SRC" and "SignedSRC". I added "compute" to the method names in order to make them more consistent with the rest of the OT API.

Concerning SRC, the new "computeSRC" method does what "SignedSRC" previously did, and "computeSquaredSRC" does what the old "SRC" did, so I think I implemented technical committee decision correctly.

For me the question is: should we maintain backwards compatibility despite this change? And if yes, then how? Should I create additional static methods "SRC" and "SignedSRC" to keep the old bad names in 1.20?

vchabri commented:
Hi all,
Thanks Joseph for this nice PR.

By the way, I have a suggestion: would it be possible to add a method "computeKendallTau" to the CorrelationAnalysis class?

This would be a nice additional feature to this class.

JPelamatti commented on lib/src/Base/Stat/CorrelationAnalysis.cxx:
```diff
@@ -26,12 +26,57 @@
+  , inputSample_(inputSample)
+  , outputSample_(outputSample)
 {
   if (outputSample.getDimension() != 1) throw InvalidDimensionException(HERE) << "Error: output sample must be 1D";
```
Could we perform a single test here? i.e., `if (outputSample.getDimension() != 1 || inputSample.getDimension() != 1 ) throw InvalidDimensionException(HERE) << "Error: both samples must be 1D";`

JPelamatti replied:
Could we perform a single test here? i.e., `if (outputSample.getDimension() != 1 || inputSample.getDimension() != 1 ) throw InvalidDimensionException(HERE) << "Error: both samples must be 1D";`

JPelamatti commented on lib/src/Base/Stat/CorrelationAnalysis.cxx:
```diff
@@ -43,63 +88,34 @@ Point CorrelationAnalysis::PearsonCorrelation(const Sample & inputSample,
+                                      const Sample & outputSample) const
 {
   const UnsignedInteger dimension = inputSample.getDimension();
   if (!(dimension >= 2)) throw InvalidDimensionException(HERE) << "Error: input sample must have dimension > 1, here dimension=" << dimension;
```
An SRC value could be computed between 2 1-dimensional samples, could it not?

josephmure replied:
Actually, there is no problem at all. I don't know why why this check was made in the first place. Removed now.

JPelamatti commented on lib/src/Base/Stat/CorrelationAnalysis.cxx:
```diff
@@ -43,63 +88,34 @@ Point CorrelationAnalysis::PearsonCorrelation(const Sample & inputSample,
+                                      const Sample & outputSample) const
 {
   const UnsignedInteger dimension = inputSample.getDimension();
   if (!(dimension >= 2)) throw InvalidDimensionException(HERE) << "Error: input sample must have dimension > 1, here dimension=" << dimension;
```
An SRC value could be computed between 2 1-dimensional samples, could it not?

mbaudin47 commented on python/doc/theory/reliability_sensitivity/ranking_src.rst:
```diff
@@ -1,70 +1,78 @@
 the literature as the “importance factor”, because of the similarity
 between this approach to linear regression and the method of cumulative
-variance quadratic which uses the term importance factor.
+variance which uses the term importance factor.
```
There are two missing points here: 
- If the physical model g is linear, then squared SRC coefficients are equal to the Sobol' indices. This is what is important to see in the squared SRC coefficient. On this point, a new pointer to the [Sobol'](https://openturns.github.io/openturns/latest/theory/reliability_sensitivity/sensitivity_sobol.html) help page should be added.
- The quality of the linear regression should be checked before estimating the SRC indices. This is because if the linear regression model is a poor fit to the data, then the SRC coefficients are useless. This is why a new pointer to the [Regression](https://openturns.github.io/openturns/latest/theory/data_analysis/linear_regression.html) help page should be added, with a suggestion: compute the R2 and check it before estimating the SRC indices.

josephmure replied:
OK, I have added this to the doc page.

mbaudin47 commented on python/doc/theory/reliability_sensitivity/ranking_src.rst:
```diff
@@ -1,70 +1,78 @@
+Standard *Rank* Regression Coefficients (SRRC) are SRC coefficients
+computed on the ranked input variables
+:math:`r\vect{X} = \left( rX^1,\ldots,rX^{n_X} \right)`
+and the ranked output variable :math:`rY`.
```
There is a missing point here: SRRC coefficient are useful when the function is monotonic but not necessarily linear. In that case, SRC coefficients cannot be used (because the function is not necessarily linear).

josephmure replied:
OK, I have added this to the doc page.

mbaudin47 commented on python/test/t_CorrelationAnalysis_std.py:
```diff
@@ -23,31 +24,30 @@
-src_normalize = ot.CorrelationAnalysis.SRC(inputSample, outputSample, True)
-print("src with normalize=", src_normalize)
+squared_src = corr_analysis.computeSquaredSRC()
+ott.assert_almost_equal(squared_src, [0.901238,0.105409], 1e-5, 0.0)
```
The same exact reference values as used in the C++ unit test should be used at this point.

josephmure replied:
Done.

mbaudin47 reviewed (changes requested):
Thank your for this good bugfix. :)

It seems that there is still some work to go from _good_ to _great_, as the current code and doc seems to lack of performance, unit test quality and cross-pointers. Perhaps a new meeting could help to clarify the API before going further.

josephmure commented on lib/test/t_CorrelationAnalysis_std.cxx:
```diff
@@ -61,34 +61,33 @@ int main(int, char *[])
-    fullprint << "signed src=" << signed_src.__str__() << std::endl;
+    // Squared SRC with normalize
+    Point squared_src_normalize(corr_analysis.computeSquaredSRC(true));
+    assert_almost_equal(squared_src_normalize, Point({0.895287,0.104713}), 1e-5, 0.0);
```
This cannot be true: by definition the sum of the normalized squared SRCs must be 1.

josephmure commented on lib/test/t_CorrelationAnalysis_std.cxx:
```diff
@@ -61,34 +61,33 @@ int main(int, char *[])
-    fullprint << "signed src=" << signed_src.__str__() << std::endl;
+    // Squared SRC with normalize
+    Point squared_src_normalize(corr_analysis.computeSquaredSRC(true));
+    assert_almost_equal(squared_src_normalize, Point({0.895287,0.104713}), 1e-5, 0.0);
```
I don't understand: this test concerns the normalized squared SRC (in this case equal to the squared SRC), not the "signed" SRC.

josephmure replied:
OK, I took the reference values for the normalized squared SRC in the test.

josephmure replied:
TODO : add the theoretical computation as a comment or a pointer to the appropriate doc page.

josephmure commented on lib/test/t_CorrelationAnalysis_std.cxx:
```diff
@@ -61,34 +61,33 @@ int main(int, char *[])
-    Point src_normalize(CorrelationAnalysis::SRC(inputSample, outputSample, true));
-    fullprint << "src with normalize=" << src_normalize.__str__() << std::endl;
+    Point squared_src = corr_analysis.computeSquaredSRC();
+    assert_almost_equal(squared_src, Point({0.901238,0.105409}), 1e-5, 0.0);
```
OK, I have increased the sample size and used the theoretical values whenever possible. I have added comments stating whether the reference value is exact or approximate.

josephmure commented on lib/test/t_CorrelationAnalysis_std.cxx:
```diff
@@ -61,34 +61,33 @@ int main(int, char *[])
     Sample inputSample(randomVector.getSample(sampleSize));
     Sample outputSample(analytical(inputSample));
```
At our meeting, we agreed to leave this to a future PR.

josephmure commented on lib/test/t_CorrelationAnalysis_std.cxx:
```diff
@@ -61,34 +61,33 @@ int main(int, char *[])
-    Point src_normalize(CorrelationAnalysis::SRC(inputSample, outputSample, true));
-    fullprint << "src with normalize=" << src_normalize.__str__() << std::endl;
+    Point squared_src = corr_analysis.computeSquaredSRC();
+    assert_almost_equal(squared_src, Point({0.901238,0.105409}), 1e-5, 0.0);
```
Done, but I had to increase the tolerance.

josephmure commented on lib/test/t_CorrelationAnalysis_std.cxx:
```diff
@@ -61,34 +61,33 @@ int main(int, char *[])
 
-    Point pearson(CorrelationAnalysis::PearsonCorrelation(inputSample, outputSample));
-    fullprint << "pearson=" << pearson.__str__() << std::endl;
+    Point pearson(corr_analysis.computePearsonCorrelation());
```
At our meeting, we agreed to leave this to a future PR.

josephmure commented on lib/src/Base/Stat/CorrelationAnalysis.cxx:
```diff
@@ -43,63 +88,34 @@ Point CorrelationAnalysis::PearsonCorrelation(const Sample & inputSample,
-    src /= varOutput;
-  else
-    throw InvalidArgumentException(HERE) << "No output variance";
+  Point src(computeSRC());
```
At our meeting, we agreed to leave this to a future PR.

vchabri replied:
Relevant remark which might have a strong impact if we consider (in the future) Shapley indices for linear models (LMG, PMVD). Minimizing the number of regressions might drastically improve the computational burden.

josephmure commented on lib/test/t_CorrelationAnalysis_std.cxx:
```diff
@@ -61,34 +61,33 @@ int main(int, char *[])
-    Point srrc(CorrelationAnalysis::SRRC(inputSample, outputSample));
-    fullprint << "srrc=" << srrc.__str__() << std::endl;
+    Point src(corr_analysis.computeSRC());
+    assert_almost_equal(src, Point({0.949336,0.324667}), 1e-5, 0.0);
```
Done.

jschueller commented on lib/src/Base/Stat/openturns/CorrelationAnalysis.hxx:
```diff
@@ -34,37 +35,92 @@ BEGIN_NAMESPACE_OPENTURNS
+
+  // computeSRC with arguments
+  Point computeSRC(const Sample & sampleX,
+                   const Sample & sampleY) const;
```
these 3 functions do not use any attributes so they can be marked static

jschueller reviewed (approved):
ok, pending static private methods

jschueller commented on lib/src/Base/Stat/openturns/CorrelationAnalysis.hxx:
```diff
@@ -107,16 +107,16 @@ public:
-  Point computeSRC(const Sample & sampleX,
-                   const Sample & sampleY) const;
+  static Point computeSRC(const Sample & sampleX,
+                          const Sample & sampleY);
```
usually they also begin with uppercase

josephmure replied:
Done.

josephmure commented:
TODO : fix API "See also" in Theory.

sofianehaddad commented on lib/src/Base/Stat/CorrelationAnalysis.cxx:
```diff
@@ -129,7 +129,7 @@ Point CorrelationAnalysis::computeSRC(const Sample & sampleX,
                                       const Sample & sampleY) const
 {
   const UnsignedInteger dimension = sampleX.getDimension();
-  if (!(dimension >= 2)) throw InvalidDimensionException(HERE) << "Error: input sample must have dimension > 1, here dimension=" << dimension;
```
You mean `sampleY`?

`sampleX` might have dimension >= 2!

josephmure replied:
I don't understand your point. I removed this check after @JPelamatti pointed out it was useless:

https://github.com/openturns/openturns/pull/2056#discussion_r886621668

sofianehaddad replied:
I did not notice the `!` in `!(dimension >= 2)`. Sorry
Everything is ok

josephmure commented on lib/src/Base/Stat/CorrelationAnalysis.cxx:
```diff
@@ -157,7 +157,7 @@ Point CorrelationAnalysis::computePCC(const Sample & sampleX,
 {
   const UnsignedInteger dimension = sampleX.getDimension();
-  if (!(dimension >= 2)) throw InvalidDimensionException(HERE) << "Error: input sample must have dimension > 1, here dimension=" << dimension;
+  //if (!(dimension >= 2)) throw InvalidDimensionException(HERE) << "Error: input sample must have dimension > 1, here dimension=" << dimension;
```
The PCC may not be very interesting in the case where `sampleX` is 1-dimensional because it is simply equal to the SRC, but the computation is correct.

vchabri commented on lib/src/Base/Stat/CorrelationAnalysis.cxx:
```diff
@@ -26,105 +26,163 @@
 BEGIN_NAMESPACE_OPENTURNS
 
-/* Compute the Pearson correlation coefficient between the component number index of the input sample and the 1D outputSample */
+/**
```
Sorry for the minor remark...
Following what @JPelamatti proposed, I was more fond of the names "firstSample / secondSample" as these metrics are "two-sample statistics". Using "sampleX / sampleY" still denotes (implicitly) an input-output relationship which might be confusing. Again, this is a tiny detail.

josephmure replied:
Done. Also removed further references to input/output variables in the `CorrelationAnalysis` API doc.

vchabri commented on python/doc/theory/reliability_sensitivity/ranking_pcc.rst:
```diff
@@ -1,51 +1,52 @@
-This method deals with analyzing the influence the random vector
+Partial Correlation Coefficients
+deal with analyzing the influence the random vector
 :math:`\vect{X} = \left( X^1,\ldots,X^{n_X} \right)` has on a random
```
@jschueller / @mbaudin47 : why do we use here a superscript-based notation for the input variables? Regarding the remainder of the documentation, one should use a subscript (e.g., X_{j}^{(i)}, for the i-th sample of the j-th variable).

jschueller replied:
I guess its a mistake, feel free to change it.

josephmure replied:
OK, done.

vchabri commented on python/doc/theory/reliability_sensitivity/ranking_src.rst:
```diff
@@ -1,70 +1,88 @@
+before estimating the SRC coefficients: if the linear regression model
+is a poor fit to the data, then the SRC coefficients are useless.
+
+Note that if there exists a map :math:`g` such that :math:`Y=g(X^1, ..., X^n)`,
```
X^n should be X_{n_X}

josephmure replied:
Changed to X^{n_X} for the time being. I'm waiting for @mbaudin47 or @jschueller's answer to your other comment https://github.com/openturns/openturns/pull/2056#discussion_r911165155 before taking further action.

vchabri reviewed (changes requested):
Good job @josephmure ! Thanks!

josephmure commented:
For some reason, continuous integration did not trigger with my latest commit but I have taken care of the super/subscript related stuff. If you want to see the result for yourself @vchabri you can try to visualize the relevant files only:

1. Pull the branch locally
2. Inside the root openturns folder, run:

```
python3 utils/docfast.py --nodiff build-onlydoc ranking_src.rst ranking_pcc.rst
```

Then you should find the resulting HTML at build-onlydoc/install/theory/reliability_sensitivity/reliability_sensitivity.html
